### DISCUSSION TITLE
Generic Type Assumption from enum key

### DISCUSSION THREAD

**Discussion by InvincibleRMC** [python/typing#1831] [Q&A] [answered]
Is there any way to type annotate so `a` is of type Test[int] instead of Test[Any]?

```python

from typing import overload, TypeVar, Literal
T = TypeVar('T')
from enum import Enum

class TypeEnum(Enum):
    INT = 0
    BOOL = 1

class Test(Generic[T]):


    @overload
    def __init__(self, foo: Literal[TypeEnum.INT], value: int | None) -> None:...

    @overload
    def __init__(self, foo: None, value: T) -> None: ...

    def __init__(self, foo, value):
        self.value = value

    @property
    def value(self) -> T:
        return self._value
    
    @value.setter
    def value(self, value: T) -> None:
        self._value = value


a = Test(TypeEnum.INT, None)
reveal_type(a)

b = Test(None, 3)
reveal_type(b)
```

**Comment by erictraut** [comment, marked as answer]
Yes, you can add a type annotation for the `self` parameter in the first overload. [Pyright playground link](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDiApikSEgMYA0UYAbmQDZgCGAJjQCqJEBqLIGgBkkMMi0YBYAFAzOUALxRuCPgIAUAIk6aAlDNCQoJAK7QsOPFACiKMzIfSKjFgGdXynrbPrvEXQBcMlAhUACSAHLySgAMwaEAQgDySUKKUACMjjLObh6cRK4w6sSk5BQA2pwAuoHxIQAC9EysbPVQbETAUAD6PaiifertoVCuRIzAAVDahTAVqDDVmjTAYGDTImIgElVedhAAdJE1NHQSJkTTi1AAPlARYKTtulAAtAB8D09XI6GHAMcoyaDBAzHY7U63T6AxgQ3Gk1W62mj1IZwuV2Ur0%2B31IQWko1GAMOQNCUN6-RQgx66gRwCRYHRjEudQJhJCdMO52ZRHS3MupMaCHAqjwcEhXSg-KItImwGxX04%2BPZIRARBgJhAKDGcsOPWlgqgDWlh3GMG2Eu60tliKlGOmnAVuN%2BbPZnP1GL5GOy0hY6QKRXUKiIfmOURoqKI%2BmkaoYEh68FU6hY0ZkACN-XN1JGaABmaOxojxxMytPRoA).

For details, refer to [this section](https://typing.readthedocs.io/en/latest/spec/constructors.html#init-method) of the typing spec.

**Reply by InvincibleRMC** [reply]
Ah that makes. Thanks so much.
